plugins/modules/gitlab_runner.py

```diff
@@ -342,7 +342,10 @@ def update_runner(self, runner, arguments):
     @param description Description of the runner
     '''
     def find_runner(self, description):
-        runners = self._runners_endpoint(as_list=False)
+        if LooseVersion(gitlab.__version__) < LooseVersion('3.6.0'):
+            runners = self._runners_endpoint(as_list=False)
+        else:
+            runners = self._runners_endpoint(iterator=True)
```

Comment:
Instead of having to do a similar comparison everywhere, does it make sense to do it once in the gitlab module_utils and store the parameter(s) needed in a dict, like `self.list_all_kwargs = ...`, and then replace this and similar calls with `runners = self._runners_endpoint(**self.list_all_kwargs)`?

Assuming of course that all these list functions behave the same and accept the same parameters.

CC @nejch

Reply:
From looking at `rg --fixed-strings '.list' plugins/modules/gitlab_*.py`:

* Several modules are just passing `all=True` (which will cause the list method to gather everything and return a list)
* One module (gitlab_project_badge) is already using `iterator=True` (but doesn't check `gitlab.__version__`)

I'd imagine most of those could be adapted to use `**list_all_kwargs` easily. At least a few of them are just grabbing item 0 from the returned list, and many others just iterate the list (and so could trivially use `iterator=True`).

Reply:
Might it be better to merge this PR as-is now (and maybe get a new 8.2 or 8.3 release out with the fix in place), then have a follow-on GitHub issue for "unifying" the kwargs across modules?

Or I could just go ahead and add that to this PR. Would I want to add a `trivial` section to the changelog fragment calling out the change?

Reply:
I'm with @felixfontein on this one - this issue is probably present with all modules that use paginated listing, it also feels like these helpers would belong in the utils IMO.

I agree with going for `iterator=True`. It also avoids the confusion/collision between python-gitlab's old `all=True` (paginate everything) vs. the `all=true` query param present in some GitLab API endpoints. Right now, when you see `all=true`, you need to check if it was intended as a python-gitlab argument or as a GitLab API parameter as it's not the same behavior.

Reply:
I suppose I (or someone) can manually audit each `.list` call in `plugins/modules/gitlab*.py`, compare against the GitLab REST API documentation, and check if any APIs take an `all` parameter...

Based on python-gitlab/python-gitlab#1458 I would _guess_ that only the Commits API takes that parameter. (Unfortunately GitLab doesn't publish an exhaustive OpenAPI spec, so confirming this would require exhaustively searching the docs, or downloading the GitLab source code and grepping.) If we can trust/assume it is the case that the Commits API is the only one which can take an `all` parameter: I've checked the python-gitlab code, and everything that would be accessing that API (a function call or the API object) is called `commits`, and that name appears nowhere in `plugins/modules/gitlab*.py`.

But that doesn't mean we can just change `all=True` to `get_all=True` across the board, because `get_all` was added in 3.7.0, but the code tries to maintain compatibility for other versions of python-gitlab.

Maybe the solution will be a helper function in module_utils, to return the necessary kwargs for a given request? It could look at the `python-gitlab` version, and decide on the right combination of `get_all`, `all`, `iterator` and `as_list`, and could pass a default `per_page` of 100.

Reply:
@felixfontein @nejch I've gone ahead and added a `list_all_kwargs` variable to the gitlab module_utils file, and made use of it in gitlab_runner. If this implementation looks okay to you, I'll work on updating other modules to use `list_all_kwargs`.